File: java/Solution746.java
package leetCode.java;

//DP O(n) O(1)
class Solution746 {
    public int minCostClimbingStairs(int[] cost) {
        int cur = 0, pre = 0;
        // i<2可以直接走到, 最多要計算到cost.length, 計算範圍是[2, n]
        for (int i = 2; i <= cost.length; ++i) {
            int temp = cur;
            cur = Math.min(cur + cost[i - 1], pre + cost[i - 2]);
            pre = temp;
        }
        return cur;
    }
}
/*       0  1  2
cost = [10,15,20] 前兩格都可以直接走到
dp[0] = 0
dp[1] = 0
dp[2] = min(dp[0] + 10, dp[1] + 15) = 10
dp[3] = min(dp[1] + 15, dp[2] + 20) = 15
...
dp[n] = min(dp[n-1] + cost[n-1], dp[n-2] + cost[n-2])
*/

File: c++/Solution746.cpp
//DP O(n) O(1)
class Solution746 {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        int one = 0, two = 0;
        for(int i=2; i<=cost.size(); ++i){//要跨過floor, 所以算到邊界外
            int tmp = one;
            one = min(one + cost[i - 1], two + cost[i - 2]);
            two = tmp;
        }
        return one;
    }
};
/* cost = [1,100,1,1], dp[i] i 表示位置

dp[0]=0
dp[1]=0
dp[2]=dp[0]+1=1
dp[3]=dp[2]+1=2
...
dp[n]=min(dp[n-1]+cost[n-1], dp[n-2]+cost[n-2])

*/